### DISCUSSION TITLE
Any parcel-compatible test runners?

### DISCUSSION THREAD

**Discussion by erikpukinskis** [parcel-bundler/parcel#7615] [Q&A] [answered]
I'm spinning up a new project, and I've already bounced from Snowpack to Parcel and on to Vite...

I was very impressed with Parcel, but the piece that seemed missing was some kind of test runner. I know I can run Jest on Babel alongside Parcel. But I'd really love to have a unified build environment for all three of dev server, production build, and tests.

With Snowpack I can share the Babel config between Jest and Snowpack. Vite has Vitest. Is there any option here for Parcel where I can get the same build in my test env as my dev/production builds?

It seems like maybe I could potentially use Karma, and just run the tests in the browser. Is that The Way™? I see there is a [karma-parcel](https://github.com/valotas/karma-parcel)odule. I'd been kind of leaning towards headless, but this could be a reason for me to just go "headful".

**Comment by devongovett** [comment, marked as answer]
On the [roadmap](https://github.com/parcel-bundler/parcel/discussions/7345#register), we have a `@parcel/register` and `@parcel/jest` section which covers this. Once implemented, this will allow reusing Parcel's transformers and resolvers in Node, or inside e.g. test frameworks.

Since you mentioned babel though, yes, you can use the same Babel config with Parcel and other tools. Or, you can just use Babel with other tools, and use Parcel without Babel as documented here: https://parceljs.org/languages/javascript/#usage-with-other-tools.

**Reply by erikpukinskis** [reply]
Thanks for the info @devongovett! That would be very exciting. I subscribed to the linked issue to get updates.

It doesn't seem like any of the next gen build tools really support what I'm looking for. There is a `vite-jest` package but I couldn't get it to work and it still seems experimental. IMO this would be a massive selling point for whichever build tool gets there first.

The four environments my JavaScript code really needs to run in are:
1) Dev server
2) Production build
3) A test environment
4) Storybook

And ideally all of them are fast, and share config! Currently no tool really hits all four. You can probably share the same babel/webpack config in all places, but it's slow. Vite in theory could do all four but vite-jest is still experimental. It is true that I can mix and match, but I would just love to not have to maintain separate configs for each different environment.

Anyway, much thanks for your work on Parcel, it gave me great vibes when I was trialing it! Apologies for mentioning a competitor, just wanted to pass along my thinking as I'm evaluating in case it is of interest!
